### PR TITLE
Turn CSVRecord into a List

### PR DESCRIPTION
**Pull request by mina86**
CSVRecord implements get(int) and size() methods so it’s already pretty
much a list.  However, because it does not implement the List interface,
users cannot take advantage of convenience methods such as subList nor
can they pass records around to methods accepting lists.

Make CSVRecord extend AbstractList so that it implements List interface.

Because AbstractList pravides iterator() implementation, this allows us
to delete said method (alongside toList() method) reducing amount of
code.

### CODE REVIEW

**Comment by garydgregory** [comment]
@mina86 ,

Thank you for your patch. We cannot break binary compatibility within the 1.x release line. That means you cannot remove public APIs for example.

Gary

**Comment by mina86** [comment]
> I wonder what others think.

Well, that’s not something I can answer…

**Comment by garydgregory** [comment]
I will ask on the dev ML for opinions.

**Comment by kinow** [comment]
This looks OK to me. It would still be an `Iterable`, as `AbstractList` also [implements it](https://docs.oracle.com/javase/7/docs/api/java/util/AbstractList.html). As the data values are not streamed, but rather kept in the internal `values[]` array, I think there is no performance issues by changing it to `AbstractList`.

I have only one concern here. Before this change, users have a complete `Iterable` object, which they can pass around (indeed not as a `List`, but they can use the parent interface or use helper methods to convert to `List`/`Collection`/etc). 

But I believe we have only `get` and `size` implemented from the `List` interface. So while users would now have a `List` object, I believe they would get `UnsupportedOperationException` for `remove`, `addAll`, `add`, `clear`, etc.

We could add documentation alerting users about potential issues of using it as a `List`, but I think this could still lead to unexpected issues reported, and probably the need to keep implementing more methods, or liaise and indoctrinate users on how to use `CSVRecord`.

For that I am -0 for this change (i.e. my code would still work as I'm using `Iterable`, and if my memory helps I won't have exceptions for a while, but if others are strongly keen to have this feature, then I'd be fine).

Hope that helps
Bruno

**Comment by kinow** [comment]
By the way, the code is really neat @mina86, with good tests. Hope to see other contributions like this 👍

**Comment by mina86** [comment]
> But I believe we have only `get` and `size` implemented from the `List` interface. So while users would now have a `List` object, I believe they would get `UnsupportedOperationException` for `remove`, `addAll`, `add`, `clear`, etc.

While that’s true, this is exactly how the interface is defined:

>> The "destructive" methods contained in this interface, that is, the methods that modify the collection on which they operate, are specified to throw UnsupportedOperationException if this collection does not support the operation.

(this particular quote is from Collection’s documentation) and I would expect no one who has any experience with Java to be surprised at that. There are multiple examples of lists which cannot be modified: `Collections.unmodifiableList` turns modifiable list into unmodifiable one, `Arrays.asList` supports `set` but does not support `add` or `remove`. Even Apache Commons has [UnmodifiableList](https://commons.apache.org/proper/commons-collections/javadocs/api-4.2/org/apache/commons/collections4/list/UnmodifiableList.html) for some reason.

File: src/changes/changes.xml [hunk 1]
```diff
@@ -43,6 +43,7 @@
       <action issue="CSV-208" type="fix" dev="ggregory" due-to="Jurrie Overgoor">Fix escape character for POSTGRESQL_TEXT and POSTGRESQL_CSV formats.</action>
       <action issue="CSV-232" type="fix" dev="ggregory" due-to="Jurrie Overgoor, Gary Gregory">Site link "Source Repository" does not work.</action>
       <action issue="CSV-234" type="add" dev="ggregory" due-to="Roberto Benedetti, Gary Gregory">Add support for java.sql.Clob.</action>
+      <action type="add" dev="mina86" due-to="Michal Nazarewicz">Turn CSVRecord into a List.</action>
```
[the remark this answers is not in the document]
Comment (mina86) [reply src/changes/changes.xml]: Yeah, I wasn’t sure what to put there. Or whether I should have edit this file at all.

[the remark this answers is not in the document]
Comment (kinow) [reply src/changes/changes.xml]: Not a problem, it's not well documented what contributors should do about `changes.xml`. You can leave this file unchanged, and a committer will add this line later with his/her ID, and the `due-to` entry as above 👍

**Comment by kinow** [comment]
I think your arguments above are good @mina86. I may be too pessimistic on my point of view, thinking that users could believe they can modify CSVRecord while parsing a CSV file and writing to another CSV file.

But as I said, I don't oppose this change 🙂 the code is good, it has tests, follows the code standards, and has a practical use. So we just need to wait for others to chime in and see what they prefer.

**Comment by KrzychuJedi** [comment]
If I may add something from developer perspective. I recently used commons-csv and this feature would be very desire change. It will allow to use stream api and simplify writing clean code. In some cases using iterator complicates writing clean code especially when business logic depends on values from cell on specific index.

**Comment by mina86** [comment]
FYI, you can do `StreamSupport.stream(record.splitterator(), false)` to convert CSVRecord into a stream though it’s less efficient in some cases (since the stream is unsized) than if CSVRecord was a List.

**Comment by garydgregory** [comment]
Note that `CSVRecord` in git master now supports() `toList()` and `stream()`. I'm not not convinced that it makes sense for a record to be an actual list.

**Comment by garydgregory** [comment]
Closing in favor of using `toList()`.

**Pull request closed without merging**